lightningd/plugin.c

```diff
@@ -256,6 +256,91 @@ static void plugin_request_queue(struct plugin_request *req)
 	io_wake(req->plugin);
 }
 
+static void plugin_log_handle(struct plugin *plugin, const jsmntok_t *paramstok)
+{
+	const jsmntok_t *msgtok, *leveltok;
+	enum log_level level;
+	msgtok = json_get_member(plugin->buffer, paramstok, "message");
+	leveltok = json_get_member(plugin->buffer, paramstok, "level");
+
+	if (!msgtok) {
+		plugin_kill(plugin, "Log notification from plugin doesn't have "
+				    "a \"message\" field");
+		return;
+	}
+
+	if (!leveltok || json_tok_streq(plugin->buffer, leveltok, "info"))
+		level = LOG_INFORM;
+	else if (json_tok_streq(plugin->buffer, leveltok, "debug"))
+		level = LOG_DBG;
+	else if (json_tok_streq(plugin->buffer, leveltok, "warn"))
+		level = LOG_UNUSUAL;
+	else if (json_tok_streq(plugin->buffer, leveltok, "error"))
+		level = LOG_BROKEN;
+	else {
+		plugin_kill(plugin,
+			    "Unknown log-level %.*s, valid values are "
+			    "\"debug\", \"info\", \"warn\", or \"error\".",
+			    json_tok_len(leveltok),
+			    json_tok_contents(plugin->buffer, leveltok));
+		return;
+	}
+
+	/* We strip the \" otherwise they'd be printed here. */
+	log_(plugin->log, level, "log: %.*s", json_tok_len(msgtok) - 2,
+	     json_tok_contents(plugin->buffer, msgtok) + 1);
+}
+
+static void plugin_notification_handle(struct plugin *plugin,
+				       const jsmntok_t *methtok,
+				       const jsmntok_t *paramstok)
+{
+	/* Dispatch incoming notifications. This is currently limited
+	 * to just a few method types, should this ever become
+	 * unwieldy we can switch to the AUTODATA construction to
+	 * register notification handlers in a variety of places. */
+	if (json_tok_streq(plugin->buffer, methtok, "log")) {
+		plugin_log_handle(plugin, paramstok);
+	} else {
+		plugin_kill(plugin, "Unknown notification method %.*s",
+			    json_tok_len(methtok),
+			    json_tok_contents(plugin->buffer, methtok));
+	}
+}
+
+static void plugin_response_handle(struct plugin *plugin,
+				   const jsmntok_t *idtok,
+				   const jsmntok_t *resulttok,
+				   const jsmntok_t *errortok)
+{
+	struct plugin_request *request;
+	u64 id;
+	/* We only send u64 ids, so if this fails it's a critical error */
+	if (!json_to_u64(plugin->buffer, idtok, &id)) {
+		plugin_kill(plugin,
+			    "JSON-RPC response \"id\"-field is not a u64");
+		return;
+	}
+
+	request = uintmap_get(&plugin->plugins->pending_requests, id);
+
+	if (!request) {
+		plugin_kill(
+		    plugin,
+		    "Received a JSON-RPC response for non-existent request");
+		return;
+	}
+
+	/* We expect the request->cb to copy if needed */
+	request->response = plugin->buffer;
```

Comment:
So the request message body is replied back in the response? This will not make the plugin display the responds or an I mistaking? I saw in line 290 you push the message to the clightning logfiles as I had expected it

Reply:
This was a refactor and not part of the notifications handling. This is when we receive a response to a call that we sent ourselves (and not a notification which doesn't have an ID). See the split in lines 389 - 432. The `request` struct contains all data that is related to this request-response cycle and will be used to dispatch the response to the caller (which might be internal or an external JSON-RPC call that we forwarded).

This is just moved into its own function from here: https://github.com/ElementsProject/lightning/blob/c497bad2e92a687e181d32adf02af05b8a96bfd9/lightningd/plugin.c#L317-L322